Implement eth_estimateGas endpoints

atchapcyp opened the pull request:
ref: https://github.com/loomnetwork/loomchain/issues/1483
- [ ] I added unit tests for any code that added
- [ ] I updated the CHANGELOG.md 
- [ ] All IP is original and not copied from another source
- [ ] I assign all copyright to Loom Network for the code in the pull request

atchapcyp committed:
Start implement eth_EstimateGas endpoint

atchapcyp committed:
evm/loomevm.go

atchapcyp committed:
add truffle test for eth_estimateGas`

atchapcyp committed:
Merge branch 'master' into eth-estimate-gas

atchapcyp committed:
eth_estimateGas truffle test

atchapcyp committed:
Merge branch 'master' into eth-estimate-gas

atchapcyp committed:
Uncomment test

enlight commented on e2e/tests/truffle/contracts/SimpleStore.sol:
```diff
@@ -0,0 +1,16 @@
+pragma solidity ^0.5.0;
+
+contract SimpleStore {
```
Call this `GasEstimateTestContract`

enlight commented on e2e/tests/truffle/test/EthFunctions.js:
```diff
@@ -144,5 +145,38 @@ contract('MyToken', async (accounts) => {
   });
 
-});
+  it('MyToken eth_estimateGas', async () => {
```
One test case is enough I think, you can remove this one and keep the one below.

enlight commented on evm/evm.go:
```diff
@@ -264,6 +264,35 @@ func (e Evm) GetStorageAt(addr loom.Address, key []byte) ([]byte, error) {
+		if e.validateTxValue && val.Cmp(common.Big0) < 0 {
+			return 0, errors.Errorf("value %v must be non negative", value)
+		}
+	}
```
Should be
```golang
if value == nil || value.Int == nil {
  val == common.Big0
} else {
  val = value.Int
  if val.Cmp(common.Big0) < 0 {
    return 0, errors.Errorf("value %v must be non negative", value)
  }
}
```

enlight commented on evm/loomevm.go:
```diff
@@ -209,6 +210,8 @@ func (lvm LoomVm) Create(caller loom.Address, code []byte, value *loom.BigUInt)
 
 func (lvm LoomVm) Call(caller, addr loom.Address, input []byte, value *loom.BigUInt) ([]byte, error) {
+	fmt.Println("PASS LOOM VM CALL")
+	defer fmt.Println("END LOOM VM CALL")
```
Remove these

enlight commented on plugin/vm.go:
```diff
@@ -188,6 +189,7 @@ func (vm *PluginVM) StaticCall(caller, addr loom.Address, input []byte) ([]byte,
 }
 
 func (vm *PluginVM) CallEVM(caller, addr loom.Address, input []byte, value *loom.BigUInt) ([]byte, error) {
+	fmt.Println("PASS PLUGIN EVM CALL")
```
Remove

enlight commented on rpc/eth/jsonrpc_conversion.go:
```diff
@@ -103,7 +103,7 @@ type JsonTxCallObject struct {
 	Value    Quantity `json:"value,omitempty"`
 	Data     Data     `json:"data,omitempty"`
-	Nonce    Quantity `json:"nonce.omitempty"`
+	Nonce    Quantity `json:"nonce,omitempty"`
```
Huh, can you submit a separate PR for this fix? I want it to go in ASAP

enlight commented on rpc/query_server.go:
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
+func (s *QueryServer) EthEstimateGas(query eth.JsonTxCallObject, block eth.BlockHeight) (resp eth.Quantity, err error) {
+	var caller loom.Address
+	if len(query.From) > 0 {
+		caller, err = eth.DecDataToAddress(s.ChainID, query.From)
```
Nope, the `from` chain ID should always be assumed to be `eth`.
Yes, `EthCall` does this wrong too, it'll be fixed in another PR I'm working on.

enlight commented on rpc/query_server.go:
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
+	vm := levm.NewLoomVm(snapshot, nil, nil, nil, true)
+	gasUsed, err := vm.EstimateGas(caller, contract, data, nil)
+	if err != nil {
+		fmt.Printf("error in call estimateGas %+v", err)
```
This error should be returned

enlight commented on rpc/query_server.go:
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
+	if err != nil {
+		fmt.Printf("error in call estimateGas %+v", err)
+	}
+	fmt.Println("query_server est. gas : ", gasUsed)
```
Remove this line

atchapcyp committed:
fix e2e test

atchapcyp committed:
remove debug message

eduardonunesp commented:
I have tested this branch together with `eth_sendRawTransaction` and the estimated gas makes sense, I'll put a test for it LoomProvider2

enlight commented on rpc/query_server.go:
```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
+	snapshot := s.StateProvider.ReadOnlyState()
+	defer snapshot.Release()
+
+	vm := levm.NewLoomVm(snapshot, nil, nil, nil, true)
```
Do we really need debug mode enabled in the EVM to estimate gas?

enlight replied:
Also, if we don't pass the account balance manager to `NewLoomVm` then any call to a payable function, or a function that does a balance check is likely to produce an inaccurate gas estimate since the code path that will be executed is likely to be quite different from the one that will be executed when ETH is available.

atchapcyp committed:
merge master

atchapcyp committed:
fix failed test

atchapcyp committed:
add truflle test for contract that use payable function

atchapcyp committed:
merge master

atchapcyp committed:
resolve conflict

enlight commented on e2e/tests/truffle/contracts/MyCoin.sol:
```diff
@@ -11,8 +11,37 @@ contract MyCoin is ERC20 {
+  function getUint() public view returns (uint ){
+      return globalUint;
+  }
+
```
Formatting is messed up in this file, it's time to add a Solidity code formatter.
Yarn install this https://github.com/duaraghav8/Ethlint and set it up in the `e2e/tests/truffle` dir.
Can copy the linter config from the `mainnet` dir in the tg-v2 repo.

atchapcyp committed:
Merge branch 'master' into eth-estimate-gas

atchapcyp committed:
Add ethlint to format solidity code

atchapcyp committed:
Format MyCoin.sol

atchapcyp committed:
Solidity Indentation fix

atchapcyp commented:
rebuild sesame

enlight commented on evm/evm.go:
```diff
@@ -264,6 +264,31 @@ func (e Evm) GetStorageAt(addr loom.Address, key []byte) ([]byte, error) {
+	} else {
+		val = value.Int
+		if val.Cmp(common.Big0) < 0 {
+			return 0, errors.Errorf("value %v must be non negetive", value)
```
`non negetive` -> `positive`

atchapcyp committed:
Merge branch 'master' into eth-estimate-gas

atchapcyp committed:
fix as PR comment

atchapcyp committed:
To address now optional

atchapcyp committed:
Merge branch 'master' into eth-estimate-gas

atchapcyp committed:
fix as pr comment

atchapcyp commented:
rebuild sesame

enlight committed:
Revert changes to yarn.lock

enlight committed:
Merge branch 'master' into eth-estimate-gas

# Conflicts:
#	e2e/tests/truffle/test/EthFunctions.js
#	e2e/tests/truffle/yarn.lock

enlight committed:
Revert auto-formatting

enlight committed:
Remove ethlint

ethlint is no longer maintained, can integrate solhint later instead.

enlight committed:
Revert yarn.lock changes (again)

enlight committed:
Support gas parameter when calling eth_estimateGas

eth_estimateGas accepts a `gas` parameter that can be used to limit the
maximum amount of gas that can be used during estimation. Note that the
gas limit that's provided is capped at the per-tx gas limit set via the
on-chain config.

Also fixed some overly lax validation of from/to address parameters,
while it's acceptable to omit them in some scenarios it's not acceptable
to ignore invalid address values.

enlight committed:
Revert some earlier changes that don't need to be in this PR

enlight reviewed (approved):
LGTM
